homeassistant/components/sonos/__init__.py

```diff
@@ -273,16 +280,19 @@ def _stop_manual_heartbeat(self, event: Event | None = None) -> None:
             self.data.hosts_heartbeat()
             self.data.hosts_heartbeat = None
 
-    async def async_add_speaker(
-        self, soco: SoCo, zone_group_state_sub: SubscriptionBase | None = None
+    async def async_add_speakers(
+        self,
+        socos: set[SoCo],
+        zgs_subscription: SubscriptionBase | None,
+        zgs_subscription_uid: str | None,
     ) -> None:
-        """Create and set up a new SonosSpeaker instance."""
+        """Create and set up new SonosSpeaker instances."""
         async with self.creation_lock:
-            if soco.uid in self.data.discovered:
-                return
-            await self.hass.async_add_executor_job(
-                self._add_speaker, soco, zone_group_state_sub
-            )
+            for soco in socos:
+                sub = None
+                if soco.uid == zgs_subscription_uid and zgs_subscription:
+                    sub = zgs_subscription
+                await self.hass.async_add_executor_job(self._add_speaker, soco, sub)
```

Comment:
Would it make sense to add them all in a single executor job to avoid the multiple context switches?

Reply:
Yes, good call 👍